### PR TITLE
Infer something for lambda without context

### LINKED ISSUE #2631: Type of lambda without context should not be inferred as `Any`
See https://github.com/python/mypy/issues/2567#issuecomment-270188135:

A `lambda` should always be at least least inferred as `Callable` (with a specific number of arguments). We know as much about it as we do about an unannotated `def`. This is really strange:
```python
def f(x): return x
g = lambda x: x
reveal_type(f)
reveal_type(g)
```
Gives
```
__tmp__.py:3: error: Revealed type is 'def (x: Any) -> Any'
__tmp__.py:4: error: Revealed type is 'Any'
```
I think they should reveal the same type.

### PR DESCRIPTION
**Pull request by elazarg**
Fix #2631

### CODE REVIEW

File: mypy/checkexpr.py [hunk 1]
```diff
@@ -1743,15 +1743,9 @@ def visit_func_expr(self, e: FuncExpr) -> Type:
...
             # No useful type context.
-            ret_type = e.expr().accept(self.chk)
-            if not e.arguments:
-                # Form 'lambda: e'; just use the inferred return type.
-                return CallableType([], [], [], ret_type, self.named_type('builtins.function'))
-            else:
-                # TODO: Consider reporting an error. However, this is fine if
-                # we are just doing the first pass in contextual type
-                # inference.
-                return AnyType()
```
Comment (elazarg) [inline mypy/checkexpr.py]: I don't know how does the shadowing work, but it seems to work (`lambda x: x` does not return the type of an outer `x`)

File: mypy/main.py [hunk 2]
```diff
@@ -480,7 +480,7 @@ def get_init_file(dir: str) -> Optional[str]:
     # These two are for backwards compatibility
     'silent_imports': bool,
     'almost_silent': bool,
-}
+}  # type: Dict[str, Any]
```
Comment (elazarg) [inline mypy/main.py]: Silence warning in line 562: `"object" not callable`

File: test-data/unit/check-inference-context.test [hunk 3]
```diff
@@ -577,12 +577,17 @@ list_a = [] # type: List[A]
...
 [case testInvalidContextForLambda]
 from typing import Callable
 f = lambda x: A() # type: Callable[[], A]
 f2 = lambda: A() # type: Callable[[A], A]
 class A: pass
 [out]
+main:2: error: Incompatible types in assignment (expression has type Callable[[Any], A], variable has type Callable[[], A])
 main:2: error: Cannot infer type of lambda
 main:3: error: Incompatible types in assignment (expression has type Callable[[], A], variable has type Callable[[A], A])
 main:3: error: Cannot infer type of lambda
```
Comment (elazarg) [inline test-data/unit/check-inference-context.test]: Perhaps we don't need this error message now?

File: mypy/checkexpr.py [hunk 4]
```diff
@@ -1743,15 +1743,9 @@ def visit_func_expr(self, e: FuncExpr) -> Type:
...
+            ret_type = self.accept(e.expr())
+            fallback = self.named_type('builtins.function')
+            return callable_type(e, fallback, ret_type)
```
Comment (elazarg) [inline mypy/checkexpr.py]: I don't know how does the shadowing work, but it seems to work (`lambda x: x` does not return the type of an outer `x`)

**Comment by gvanrossum** [comment]
With this PR in place, in PY2 mode, the following example (simplified from some real-world code) gives an error:
```python
def f(x):
    return x.strip()
g = lambda line: map(f, line.split(':'))
lines = ['']
dict([g(line) for line in lines if len(g(line)) == 2])  # <-- here
```
The error is:
```
__tmp__.py:5: error: List comprehension has incompatible type List[List[Any]]
```

**Comment by gvanrossum** [comment]
I keep going back and forth over what's going on in that example. You'd think the type of `g(line)` should be `List[str]` because `line` is a `str`. Then the problem is that `dict([['']])` is wrongly typed. (The code intends to make a list of *pairs* of strings, but it uses `List[str]` instead of `Tuple[str, str]`, and then the argument to `dict()` is wrong.)

But the type of `g(line)` is actually `List[Any]` because it is equivalent to
```python
def g(line: Any) -> XXX:
    return map(f, line.split(':'))
```
where `XXX` is the type of the return expression, i.e. `map(f, line.split(':'))` -- which is `List[Any]` because at this point `line` is only known to have the type `Any`. This is shown by checking
```python
g = lambda line: [line]
reveal_type(g(''))  # E: Revealed type is 'builtins.list[Any]'
```
Without this PR the type would just be `Any`, shutting up the type checker.

So on the one hand I conclude that the original code (from which I boiled this down) would have a type error even if we had coerced the type of `g` to `Callable[[str], List[str]]`. So the PR actually found a type error that was hidden before because any `lambda` would be inferred as type `Any` without a type context.

OTOH maybe we could do better by constructing a *generic* function from the lambda, so that the types here would correspond:
```python
g = lambda line: [line]  # corresponds to Callable[[T], List[T]]
```

**Comment by elazarg** [comment]
Inferring generic lambda will be more precise, which will probably catch more errors; it won't silence anything that's not silenced now. So this suggestion is independent from the original example.

One potential problem with inferring generic lambda is that it might be too strict - it will be hard to specify `Any` for some arguments selectively. Why don't we infer generic function parameters?

**Comment by JukkaL** [comment]
Inferring generic types for lambdas (when there's no type context) would require some heavy machinery (unless we special case some simple cases only). We only infer lambda argument types in trivial cases where the type context has a callable type, and I don't think that anything substantially more complicated is worth the effort right now.

Instead of inferring `Any` as the argument type when there is no help from type context, it might be more consistent to give up and ask help from the user, similar to how we don't infer `List[Any]` for `[]` when we can't determine the item type. However, annotating lambdas using `# type: Callable[<...>]` is much less obvious than annotating lists, and lambdas without type context don't seem to be very common, so falling back to `Any` doesn't sound very bad. Maybe we should at least give an error if there is an implicit `Any` argument type when using `--disallow-untyped-defs`? We can create a follow-up issue for that.

**Comment by gvanrossum** [comment]
Remember, the lambda syntax does not support type annotations.

Brainstorming a little more, maybe in
```python
f = lambda x: ...
foo(f)
```
we could just save the whole lambda as the value/type of `f`, and type-check the use of `f` as if it was the original lambda? The places where I've seen code like this looked like the only reason to introduce the variable was to break a long line.

**Comment by gvanrossum** [comment]
(I'm also okay with merging this and maybe iterating later.)

**Comment by gvanrossum** [comment]
Hm, I found another regression in a different codebase.
```python
import re
p = re.compile(r".*")
print(p.sub(lambda *a, **k: "ho", "abc"))
```
This was fine before, but now says
```
__tmp__.py:3: error: Cannot infer type of lambda
```

**Comment by elazarg** [comment]
But the following fails with or without this PR:
```python
from typing import Callable
def f(t: Callable[[str], str]) -> str: ''
f(lambda *_: "")  # E: Cannot infer type of lambda
```
(Since `callable_ctx.arg_kinds != arg_kinds`)

It feels like it is unrelated to this PR, since this error happens before the decision to infer "something". I will try to find the cause anyway; maybe I'm wrong.

**Comment by elazarg** [comment]
(Sorry: of course this is related to this PR since there's no error without it)

**Comment by elazarg** [comment]
Ok, I get it: the error "occurs" without this PR too, but returning `AnyType()` has the implicit meaning of silencing this error. We cannot infer the type of the lambda even though it has context, but we do not notify the user.

**Comment by gvanrossum** [comment]
Maybe this case is the reason the original code had a TODO at the `return
AnyType()` suggesting to make it an error?

**Comment by elazarg** [comment]
If so, I should move and rephrase the TODO.
(Also, I am not sure this specific case is an example of where we can't reasonably infer the type).

**Comment by gvanrossum** [comment]
The real-world examples I found all basically just ignore any arguments
they receive. So it should be fine to infer Callable[..., X] where X is
determined by the expression only.

**Comment by elazarg** [comment]
I think this specific case can be treated separately. For now there's a bandage on it.

(I am also not sure about my previous conclusion that `Any` silences the errors - it could be something to do with the overload in the example)

**Comment by gvanrossum** [comment]
So I have one more real-world issue with this. Consider this example:
```python
from typing import Callable
def f(g = lambda *a, **k: None):
    # type: (Callable[..., None]) -> None
    g(1, 2)
```
This gives an error:
```
__tmp__.py:2: error: Incompatible types in assignment (expression has type Callable[[StarArg(Any), KwArg(Any)], None], variable has type Callable[..., None])
```
The problem is that I don't even know how to silence this -- you can't put `# type: ignore` on a `# type comment` and I don't know how to make the arguments more general than `...`. The only solution would be to use plain `Callable` but I would prefer to keep the `None` return.

Thoughts?

**Comment by elazarg** [comment]
Turns out it's not about the the parameters but about the return type. When you say `Callable[..., None]` mypy understand `Callable[..., Void]` but the lambda was inferred to return `builtins.None`.

There's actually a [test](https://github.com/python/mypy/commit/741543fc51a3218a88b98d1da7d3c03d2eab4bc5) for this behavior, which fixed #1425.
Changing the inference to return Void does not resurface #1425. It does fail this test:
```
[case testLambdaReturningNone]
f = lambda: None
x = f()
```
Yielding `error: Function does not return a value`. I am not sure that this error is not needed - this code is obviously wrong, and in more complex situations I would expect the lambda to have context. (Thinking about it, the original problem should not have occurred at all, since it does have context; we shouldn't be inferring blindly in the first place).

In other words, for this PR to work we need to decide whether `lambda: None` is a procedure, or whether a function that returns `builtins.None` is compatible with `Callable[..., Void]`.

I would expect `Callable[..., T]` to be compatible with `Callable[..., Void]` for all T - in a way, `Void` is the top element - but there are explicit checks to avoid `None <: Void` and `UninhabitedType <: Void` even though both seem fine to me.  I am probably missing something, but if it's about catching erroneous `return`s even though they are technically sound, then IMHO these checks should not be part of the subtyping relation but an explicit special-casing somewhere else.

And maybe it's not worth the trouble.

**Comment by gvanrossum** [comment]
Thanks for getting to the bottom of this. Your analysis sounds right, but I don't know what we should do about it either. Maybe @JukkaL has an opinion? I'll also ask @ddefisher.

**Comment by gvanrossum** [comment]
We think this will also break some things in our codebases; I'll report back with results in a little while.

**Comment by gvanrossum** [comment]
Actually it didn't break anything (and of course it fixed the example I gave). So I'm going to merge this.

**Comment by gvanrossum** [comment]
There may be a follow-up issue -- do we still need the "Cannot infer type of lambda" error?

**Comment by elazarg** [comment]
I have no idea.

**Comment by gvanrossum** [comment]
Hm, actually the issue in https://github.com/python/mypy/pull/2634#issuecomment-271392350 is still there, so I filed #2764 for that.
